### DISCUSSION TITLE
Supporting multiple docker versions for WP (release) images

### DISCUSSION THREAD

**Discussion by pat-s** [woodpecker-ci/woodpecker#2453] [General]
As initially discussed in https://github.com/woodpecker-ci/woodpecker/pull/1792, it might be that WP *should* support multiple docker versions for its releases. Otherwise various OS might not be compatible as a host OS for WP (see https://github.com/woodpecker-ci/woodpecker/issues/1786).

If the docker version used by WP is newer than the one the host OS uses then running WP doesn't work. 
Some distros take their time to catch up and staying several years behind the current docker releases is also not a nice option.

Hence release images which use different docker versions might be needed. Doing so for `next` might be an overkill though.

Opinions?

**Comment by qwerty287** [comment]
I don't like this, it requires much more maintenance without real improvement.
Instead, I'd rather stay with docker 20.x and update to the latest version once the most important distros provide newer versions.

**Reply by pat-s** [reply]
Valid point. I also think that compatibility comes first here even though we should aim at staying up-to-date with deps. But it's unclear how much builds would profit from new major docker versions in the end. 

We should at least stay up-to-date with the most recent patch releases for any version.

I'll add a list where all the major distros currently stand at. Maybe we can even pin this as an issue for overview purposes?

**Comment by pat-s** [comment]
Date: (2023-09)

- Suse 15.5: **23.x**
- Ubuntu 20.04.: **24.x**
- Ubuntu 22.04.3: **24.x**
- Debian 12: **20.10.24**
- Alma 8 & 9: will always get latest docker as docker must be installed from external sources
- Amazonlinux: **20.10.25**

**Reply by anbraten** [reply]
That's quite a few already

**Reply by qwerty287** [reply]
> Debian 12: 20.10.24

Waiting for this means waiting until 2025 because debian won't update its stable branch and next release is debian 13 which will be released in 2025.
However, Debian is a pretty important system for servers so I'm not sure if we should break it.

**Reply by pat-s** [reply]
Ubuntu just caught up in the last few weeks. Same for Suse. 

Debian might issue a point release at some point? I can hardly believe they stick with 20.x until 2025? I mean they would be like 4 major versions behind by then...

**Reply by qwerty287** [reply]
Debian does not care about this usually, only security fixes are pushed. E.g. Gnome desktop is 3.38 in Debian 11 (which was the latest Debian release until June this year). They directly bumped it to 43 in Debian 12 which is only that new because Debian 12 was released 3 months ago. Gnome 3.38 was released in Sep 2020.

**Reply by xoxys** [reply]
Docker 20.10 will become EOL in ~ 3 months. I guess there will be no security updates for the go packages for EOL versions. Even if Debian backports security fixes for their packages on its own, I don't think that's a proper solution for Woodpecker and the used docker go packages.

**Reply by qwerty287** [reply]
I think we should wait with the upgrade until the EOL and update then, no matter which version Debian is using then.

**Reply by xoxys** [reply]
Sounds like a plan 👍

**Reply by pat-s** [reply]
Sounds good to me as well. It's then a Debian issue and we shouldn't support EOL docker engines in WP.
Due to Suse being stuck on 23.x for now, we should only upgrade to this version then.

Amazonlinux will most likely do a major update at some point to not be stuck with 20.x.
